# Generator Expressions vs List Comprehension 
Comparaison between the space complexity of Generator Expressions vs List Comprehension 

In [1]:
from sys import getsizeof 
# calculates the space consumption 
# of variables

In [14]:
# Initialize variables
l = [ i for i in range(10)]
g = ( i for i in range(10))

In [15]:
print("list: " + str(getsizeof(l)))
print("list: " + str(getsizeof(g)))

list: 184
list: 112


We already have higher space consumption from list

Now lets increase the size

In [18]:
# Increase the size
l = [ i for i in range(10000)]
g = ( i for i in range(10000))

In [19]:
print("list: " + str(getsizeof(l)))
print("list: " + str(getsizeof(g)))

list: 85176
list: 112


Notice how when we increase in the size of the list, the space increases linearly 
while with generator it is always 112

In [20]:
# library to count infinitly
from itertools import count


In [23]:
# initialize l to be infinitly large
l = [ i for i in count(1)]


KeyboardInterrupt: 

never stops untill interrupted

In [27]:
g = ( i for i in count(1))

getsizeof(g)

112

we notice that the size never changes. This proves that the generator only saves one element at a time